File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next
        
        targetNode = length - n + 1
        count = 0
        curr = head
        prev = None
        while curr:
            count += 1
            if count == targetNode:
                if prev == None:
                    return curr.next
                prev.next = curr.next
                break
            prev = curr
            curr = curr.next
        
        return head

File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        if (head->next == nullptr) return;
        ListNode* lastNodeFirstHalf = nullptr;
        ListNode* slow = head; 
        ListNode* fast = head;
        while (fast && fast->next) {
            fast = fast->next->next;
            lastNodeFirstHalf = slow;
            slow = slow->next;
        }
        // If odd length, fast landed on the last node (not null)
        // so the middle node belongs to the first half
        if (fast != nullptr) {
            lastNodeFirstHalf = slow;
            slow = slow->next;
        }
        // Rever the second half
        lastNodeFirstHalf->next = nullptr;
        ListNode* prev = nullptr;
        ListNode* curr = slow;

        while (curr) {
            ListNode* currNext = curr->next;
            curr->next = prev;
            prev = curr;
            curr = currNext;
        }
        ListNode* firstHalf = head;
        ListNode* secondHalf = prev;
        while (firstHalf != nullptr and secondHalf != nullptr) {
            ListNode* nextSecondHalf = secondHalf->next;
            ListNode* nextFirstHalf = firstHalf->next;

            firstHalf->next = secondHalf;
            secondHalf->next = nextFirstHalf;

            firstHalf = nextFirstHalf;
            secondHalf = nextSecondHalf;
        }
    }
};
